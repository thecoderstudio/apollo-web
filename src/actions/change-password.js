export const PROMPTED_PASSWORD_CHANGE = 'PROMPTED_PASSWORD_CHANGE';

export function promptedPasswordChange() {
  return {
    type: PROMPTED_PASSWORD_CHANGE
  };
}
