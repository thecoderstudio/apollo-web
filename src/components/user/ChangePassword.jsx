import React from 'react';
import { connect } from 'react-redux';
import { Formik } from 'formik';
import axios from 'axios';
import styled from 'styled-components';
import Button from '../buttons/Button';
import OutlinedButton from '../buttons/OutlinedButton';
import Input from '../../components/Input';
import { logout } from '../../util/auth';
import { changePasswordSchema } from '../../validation/user.js';
import { fetchCurrentUser } from '../../util/user';
import { StatusCodes } from 'http-status-codes';
import { handleHTTPResponse } from '../../actions/error';
import { parseHTTPErrors } from '../../util/parser';

const Form = styled.form`
  width: 100%;
`;

const StyledButton = styled(Button)`
  width: 100%;
`;

const StyledOutlinedButton = styled(OutlinedButton)`
  margin-top: 16px;
  width: 100%;
`;

class ChangePassword extends React.PureComponent {
  constructor(props) {
    super(props);
    this.changePassword = this.changePassword.bind(this);
    this.getDisabledButtonState = this.getDisabledButtonState.bind(this);
  }

  changePassword(values, { setErrors }) {
  axios.patch(
    `${process.env.APOLLO_HTTP_URL}user/me`,
      {
        'old_password': values['oldPassword'],
        'password': values['password'],
        'password_confirm': values['passwordConfirm']
      },
    { withCredentials: true }
  )
    .then(_ => {
      fetchCurrentUser();
    })
    .catch(error => {
      handleHTTPResponse(error.response, true, true);
      if (error.response.status === StatusCodes.BAD_REQUEST) {
        setErrors(parseHTTPErrors(error.response.data, {
          old_password: 'oldPassword', password_confirm: 'passwordConfirm'
        }));
      }
    });
  }

  getDisabledButtonState(values){
    return Object.values(values).some((value) => value === '');
  }

  render() {
    return (
      <Formik
        initialValues={{ password: '', passwordConfirm: '', oldPassword: ''}}
        validationSchema={changePasswordSchema}
        validateOnChange={false}
        onSubmit={this.changePassword}>
        {({ values, errors, handleChange, handleSubmit }) => (
        <Form onSubmit={handleSubmit}>
          <Input
            name='oldPassword'
            placeholder='Current password'
            type='password'
            value={values.oldPassword}
            error={errors.oldPassword}
            onChange={handleChange}
          />
          <Input
            name='password'
            type='password'
            placeholder='New password'
            value={values.password}
            error={errors.password}
            onChange={handleChange}
          />
          <Input
            name='passwordConfirm'
            type='password'
            placeholder='Confirm new password'
            value={values.passwordConfirm}
            error={errors.passwordConfirm}
            onChange={handleChange}
          />
          <StyledButton disabled={this.getDisabledButtonState(values) }>Change password</StyledButton>
          <StyledOutlinedButton id='logoutButton' onClick={logout}>Cancel and logout</StyledOutlinedButton>
        </Form>
        )}
      </Formik>
    );
  }
}

export default connect(
  state => ({ currentUser: state.currentUser })
)(ChangePassword);
