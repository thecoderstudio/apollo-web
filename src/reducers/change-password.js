import { prompted_PASSWORD_CHANGE } from '../actions/change-password';

export default function changePasswordReducer(state=false, action) {
  switch(action.type) {
    case prompted_PASSWORD_CHANGE:
      return true
    default: return state
  }
}
