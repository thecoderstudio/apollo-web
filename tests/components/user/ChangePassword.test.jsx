import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import waitForExpect from 'wait-for-expect';
import axios from 'axios';
import { StatusCodes } from 'http-status-codes';
import { prompedPasswordChange } from '../../../src/actions/change-password.js';
import ChangePassword from '../../../src/components/user/ChangePassword';
 
const mockStore = configureStore([]);
jest.mock('axios');

function getComponent(store) {
  return renderer.create(
    <Provider store={store}>
      <ChangePassword />
    </Provider>
  );
}

function submitForm(root, oldPassword, password, passwordConfirm) {
  const oldPasswordInput = root.findByProps({ placeholder: 'Current password' });
  const passwordInput = root.findByProps({ placeholder: 'New password' });
  const passwordConfirmInput = root.findByProps({ placeholder: 'Confirm new password' });
  const form = root.findByType('form');

  oldPasswordInput.props.onChange({ currentTarget: {
    name: 'oldPassword',
    value: oldPassword
  }});
  
  passwordInput.props.onChange({ currentTarget: {
    name: 'password',
    value: password 
  }});

  passwordConfirmInput.props.onChange({ currentTarget: {
    name: 'passwordConfirm',
    value: passwordConfirm 
  }});

  console.log("SUBMITTING");
  form.props.onSubmit();
}

describe("change password", () => {
  let store;
  let spy;

  beforeEach(() => {
    store = mockStore({
      currentUser: {
        id: '1',
        has_changed_initial_password: false
      }
    });
    process.env = {
      APOLLO_HTTP_URL: 'http://localhost:1234/'
    };
    spy = jest.spyOn(store, 'dispatch');
  });

  it("renders correctly", () => {
    expect(getComponent(store).toJSON()).toMatchSnapshot();
  });

  it("skips changing password", async () => {
    const component = getComponent(store);
    const root = component.root;

    root.findAllByType('button')[1].props.onClick();
    await waitForExpect(() => {
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(prompedPasswordChange());
      expect(axios.put).not.toHaveBeenCalled();
    });
  });

  it("validates input", async () => {
    const component = getComponent(store);
    const root = component.root;

    submitForm(root, '', 'password', 'password');
    submitForm(root, 'oldPassword', 'passw', 'passw');
    submitForm(root, 'oldpassword', 'password1', 'password2');

    await waitForExpect(() => {
      expect(axios.put).not.toHaveBeenCalled();
    });
  });

  it('put successfully', async () => {
    const component = getComponent(store);
    const root = component.root;
    
    axios.put.mockResolvedValue({
      status: StatusCodes.OK
    });
    
    submitForm(root, 'oldpassword', 'password', 'password');

    await waitForExpect(() => { 
      expect(axios.put).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledWith(prompedPasswordChange());
    })
  });

  it('handles unsuccessful put', async () => {
    const component = getComponent(store);
    const root = component.root;
    
    axios.put.mockImplementationOnce(() => Promise.reject({
      response: {
        status: StatusCodes.BAD_REQUEST,
        statusText: "Bad request",
        data: {}
      }
    }));
    
    submitForm(root, 'oldpassword', 'password', 'password');


    axios.put.mockImplementationOnce(() => Promise.reject({
      response: {
        status: StatusCodes.INTERNAL_SERVER_ERROR,
        statusText: "Something went wrong",
        data: {}
      }
    }));
    
    submitForm(root, 'oldpassword', 'password', 'password');

    await waitForExpect(() => {
      expect(axios.put).toHaveBeenCalledTimes(2);
      expect(spy).not.toHaveBeenCalled();
    });
  });
});
