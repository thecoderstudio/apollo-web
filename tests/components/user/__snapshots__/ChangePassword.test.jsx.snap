// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change password renders correctly 1`] = `
.c3 {
  max-height: 50px;
  padding: 12px;
  border: none;
  border-radius: 5px;
  color: white;
  font-family: 'B612',sans-serif;
  font-size: 1rem;
  font-weight: 700;
  cursor: pointer;
}

.c3:hover {
  opacity: 0.90;
}

.c3:focus {
  outline: none;
}

.c3:active {
  opacity: 0.80;
}

.c3:disabled:hover {
  opacity: 1;
  cursor: auto;
}

.c1 {
  height: 80px;
}

.c2 {
  min-height: 50px;
  width: 100%;
  border: none;
  border-radius: 5px;
  font-family: 'Libre Franklin',sans-serif;
  font-size: 1rem;
  padding-left: 10px;
  opacity: 0.9;
  border: 1px solid transparent;
}

.c2:focus {
  outline: none border:1px solid;
}

.c0 {
  width: 100%;
}

.c4 {
  width: 100%;
  margin-top: 15px;
}

<form
  className="c0"
  onSubmit={[Function]}
>
  <div
    className="c1"
  >
    <input
      className="c2"
      name="oldPassword"
      onChange={[Function]}
      placeholder="Current password"
      type="password"
    />
  </div>
  <div
    className="c1"
  >
    <input
      className="c2"
      name="password"
      onChange={[Function]}
      placeholder="New password"
      type="password"
      value=""
    />
  </div>
  <div
    className="c1"
  >
    <input
      className="c2"
      name="passwordConfirm"
      onChange={[Function]}
      placeholder="Confirm new password"
      type="password"
      value=""
    />
  </div>
  <button
    className="c3 c4"
    disabled={true}
  >
    Change password
  </button>
</form>
`;
